### PR TITLE
[#49] [#57] Multi-cloud support

### LINKED ISSUE #49: GCP/GKE deploy optimization
Main points:
- move all provider definitions to upper level of terraform modules to avoid duplicates and interceptions and also for more comfortable version control
- get rid of unused variables for AWS provider (switch to using environment variables)

### LINKED ISSUE #57: Adopt deploy to multiple cloud providers
In order to create/destroy clusters in different clouds (GCP/Azure/AWS) from corresponding terraform envs there should be changed:
- `tf_runner` wrapper script
- cluster profile variables and its template
- Jenkins pipelines

### PR DESCRIPTION
**Pull request by maxkochubey**
-  All provider definitions to upper level of terraform modules to avoid duplicates and interceptions and also for more comfortable version control
- Terraform docker image now includes CLI tools for Azure and AWS.
- Extra variables definitions for specific cloud providers have been removed from modules
- `tf_runner` wrapper script was taken from branch `feat/985-EKS` and rewritten
- `k8s/nginx-ingress` terraform module was taken from branch `feat/985-EKS` and rewritten

This PR fixes #49. Fixes #57.
It should be merged only after merge of corresponding branch in `legion-profiles` repo.

### CODE REVIEW

File: terraform/modules/helm_init/variables.tf [hunk 1]
```diff
@@ -1,19 +1,11 @@
...
 variable "legion_helm_repo" {
   description = "Legion helm repo"
 }
 
 variable "istio_helm_repo" {
   description = "Istio helm repo"
 }
+
+variable "tiller_image" {
+  default = "gcr.io/kubernetes-helm/tiller:v2.14.0"
```
Comment (mcdoker18) [inline terraform/modules/helm_init/variables.tf]: You upgraded the helm version in this PR then you should upgrade the tiller version too.

File: containers/terraform/Dockerfile [hunk 2]
```diff
@@ -37,15 +37,21 @@ RUN curl -fsSLO https://kubernetes-helm.storage.googleapis.com/helm-${HELM_VERSI
...
+
 # Install Hiera
-RUN curl https://bootstrap.pypa.io/get-pip.py | python3 && \
+RUN curl -sL https://bootstrap.pypa.io/get-pip.py | python3 && \
     pip3 install pyyaml==5.1.1 && \
     gem install hiera hiera-eyaml
 
-
+# Install awscli
+RUN pip3 install awscli
```
Comment (mcdoker18) [inline containers/terraform/Dockerfile]: What do you think to add awscli installation to the `pip3 install pyyaml==5.1.1 && \` line.

File: pipelines/legionPipeline.groovy [hunk 3]
```diff
@@ -13,30 +13,30 @@ def createGCPCluster() {
         file(credentialsId: "${env.hieraPrivatePKCSKey}", variable: 'PrivatePkcsKey')]) {
             withCredentials([
             file(credentialsId: "${env.hieraPublicPKCSKey}", variable: 'PublicPkcsKey')]) {
```
Comment (mcdoker18) [inline pipelines/legionPipeline.groovy:15]: Maybe we rename the functions as well? I think the PR name suits for this :)
I mean remove the `GCP` prefixes.

File: terraform/env_types/gcp/gke/gke_create/backend.tf [hunk 4]
```diff
@@ -1,11 +1,3 @@
...
-  }
-}
-
-# terraform {
-#   backend "local" {
-#     path = "../../../../_tfstate/state.tfstate"
-#   }
-# }
+  backend "gcs" {}
+}
```
Comment (mcdoker18) [inline terraform/env_types/gcp/gke/gke_create/backend.tf:3]: Why did you delete the prefixes?

File: terraform/modules/k8s/nginx-ingress/variables.tf [hunk 5]
```diff
@@ -1,37 +1,71 @@
...
-}
-
 variable "root_domain" {
+  default     = ""
   description = "Legion cluster root domain"
 }
 
 variable "allowed_ips" {
   type        = list(string)
+  default     = ["0.0.0.0/0"]
```
Comment (mcdoker18) [inline terraform/modules/k8s/nginx-ingress/variables.tf]: Are you sure that it is a good default value?

File: terraform/modules/k8s/nginx-ingress/gcp.tf [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+# Remove default nginx-ingress fw rules created by controller
+resource "null_resource" "ingress_fw_cleanup" {
+  count = local.gcp_resource_count
+  triggers = {
+    build_number = timestamp()
+  }
+  provisioner "local-exec" {
+    command = "gcloud compute firewall-rules list --filter='name:k8s-fw- AND network:${var.network_name}' --format='value(name)' --project='${var.project_id}'| while read i; do if (gcloud compute firewall-rules describe --project='${var.project_id}' $i |grep -q 'kube-system/nginx-ingress'); then gcloud compute firewall-rules delete $i --quiet; fi; done"
+  }
+  #  depends_on = [helm_release.nginx-ingress]
```
Comment (mcdoker18) [inline terraform/modules/k8s/nginx-ingress/gcp.tf]: Why did you commend out it?

File: tools/tf_runner.sh [hunk 7]
```diff
@@ -1,28 +1,27 @@
...
+	if [[ $# == 0 ]]; then
+		echo "ERROR: Options not specified! Use -h for help!"
 		exit 1
 	fi
 
-	while test $# -gt 0
-	do
-        case "$1" in
-            -h|--help)
+	while [[ $# > 0 ]];	do
```
Comment (mcdoker18) [inline tools/tf_runner.sh]: `qt`?

File: tools/tf_runner.sh [hunk 8]
```diff
@@ -33,124 +32,244 @@ function ReadArguments() {
...
 		echo $result
 	fi
 }
 
+function IngressTFCrutch() {
+	for file in $(find $1 -type f -name "*.tf" ! \( \
+	-name "main.tf" -o \
+	-name "versions.tf" -o \
+	-name "variables.tf" -o \
+	-name "$(echo $(GetParam 'cluster_type') | awk -F\/ '{print $1}').tf" \)); do
```
Comment (mcdoker18) [inline tools/tf_runner.sh]: `echo $(GetParam 'cluster_type')` -> `GetParam 'cluster_type'`

**Review by mcdoker18 — CHANGES_REQUESTED** [review]

File: tools/tf_runner.sh (hunk 7 above)
Comment (maxkochubey) [reply tools/tf_runner.sh]: Fixed in 6675214, thanks!

File: terraform/modules/helm_init/variables.tf (hunk 1 above)
Comment (maxkochubey) [reply terraform/modules/helm_init/variables.tf]: Fixed in 0121b18, thanks!

File: tools/tf_runner.sh (hunk 8 above)
Comment (maxkochubey) [reply tools/tf_runner.sh]: Fixed in 6d71952

File: terraform/env_types/gcp/gke/gke_create/backend.tf (hunk 4 above)
Comment (maxkochubey) [reply terraform/env_types/gcp/gke/gke_create/backend.tf:3]: Entire backend configuration moved to [wrapper script](https://github.com/legion-platform/legion-infrastructure/blob/feat/57-multicloud/tools/tf_runner.sh#L105).

File: containers/terraform/Dockerfile (hunk 2 above)
Comment (maxkochubey) [reply containers/terraform/Dockerfile]: Good idea, fixed by bee1f79.

File: terraform/modules/k8s/nginx-ingress/variables.tf (hunk 5 above)
Comment (maxkochubey) [reply terraform/modules/k8s/nginx-ingress/variables.tf]: You're right - removed it at all in a94586b.

File: terraform/modules/k8s/nginx-ingress/gcp.tf (hunk 6 above)
Comment (maxkochubey) [reply terraform/modules/k8s/nginx-ingress/gcp.tf]: Looks like mistake - fixed in a94586b.

File: terraform/env_types/gcp/gke/gke_create/provider.tf [hunk 9]
```diff
@@ -0,0 +1,33 @@
+provider "google" {
```
Comment (karbyshevds) [inline terraform/env_types/gcp/gke/gke_create/provider.tf:1]: Please do not declare both `google` and `google-beta` providers since we are not using different API versions.

File: terraform/modules/k8s/nginx-ingress/aws.tf [hunk 10]
```diff
@@ -0,0 +1,68 @@
+# Data
+data "aws_vpc" "default" {
```
Comment (karbyshevds) [inline terraform/modules/k8s/nginx-ingress/aws.tf]: Looks like this file is out of scope of current PR.

File: pipelines/legionPipeline.groovy (hunk 3 above)
Comment (maxkochubey) [reply pipelines/legionPipeline.groovy:15]: Functions are renamed and also slightly adopted to accept arguments for further multiple cloud parameters.

**Review by karbyshevds — CHANGES_REQUESTED** [review]

File: terraform/modules/k8s/nginx-ingress/aws.tf (hunk 10 above)
Comment (maxkochubey) [reply terraform/modules/k8s/nginx-ingress/aws.tf]: You're right. Removed in 195a61e.

**Review by karbyshevds — APPROVED** [review]

**Review by mcdoker18 — APPROVED** [review]
